File: Python/LastStoneWeightII.py
## Leetcode Problem - 1049

import math

def lastStoneWeightII(stones):

    dp = {}
    
    stoneSum = sum(stones)
    target = math.ceil(stoneSum/2)     

    def dfs(i,total):
        if total>=target or i == len(stones):
            return abs(total-(stoneSum-total))
        
        if (i,total) in dp:
            return dp[(i,total)]
        
        dp[(i,total)] = min(dfs(i+1,total),dfs(i+1,total+stones[i]))

        return dp[(i,total)]
    
    return dfs(0,0)



stones = [2,7,4,1,8,1]
print(lastStoneWeightII(stones))

File: Javascript/LastStoneWeightII.js
//Leetcode Problem - 1049

var lastStoneWeightII = function(stones){
    const stoneSum = stones.reduce((acc, stone) => acc + stone, 0);
    const target = Math.ceil(stoneSum / 2);
    const n = stones.length;

    const dp = new Array(n + 1).fill(null).map(() => new Array(target + 1).fill(0));

    for (let i = 1; i <= n; i++) {
        for (let j = 0; j <= target; j++) {
        if (stones[i - 1] <= j) {
            dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - stones[i - 1]] + stones[i - 1]);
        } else {
            dp[i][j] = dp[i - 1][j];
        }
        }
    }

    const subsetSum = dp[n][target];
    return Math.abs(stoneSum - 2 * subsetSum);
}


stones = [2,7,4,1,8,1]
console.log(lastStoneWeightII(stones))